x-pack/plugins/security_solution/server/lib/detection_engine/signals/search_after_bulk_create.ts

```diff
@@ -53,177 +58,196 @@ export const searchAfterAndBulkCreate = async ({
-          createSearchAfterReturnType({
-            searchAfterTimes: [searchDuration],
-            errors: searchErrors,
-          }),
-        ]);
-        // determine if there are any candidate signals to be processed
-        const totalHits = createTotalHitsFromSearchResult({ searchResult });
-        logger.debug(buildRuleMessage(`totalHits: ${totalHits}`));
-        logger.debug(
-          buildRuleMessage(`searchResult.hit.hits.length: ${searchResult.hits.hits.length}`)
-        );
-
-        // search results yielded zero hits so exit
-        // with search_after, these two values can be different when
-        // searching with the last sortId of a consecutive search_after
-        // yields zero hits, but there were hits using the previous
-        // sortIds.
-        // e.g. totalHits was 156, index 50 of 100 results, do another search-after
-        // this time with a new sortId, index 22 of the remaining 56, get another sortId
-        // search with that sortId, total is still 156 but the hits.hits array is empty.
-        if (totalHits === 0 || searchResult.hits.hits.length === 0) {
-          logger.debug(
-            buildRuleMessage(
-              `${
-                totalHits === 0 ? 'totalHits' : 'searchResult.hits.hits.length'
-              } was 0, exiting and moving on to next tuple`
-            )
-          );
-          break;
-        }
+  if (timestamps == null || timestamps.length === 0) {
+    logger.error(
+      buildRuleMessage(
+        `No indices contained timestamp fields: ${JSON.stringify(timestampsToSort, null, 2)}`
+      )
+    );
+    throw Error(
+      `No indices contained timestamp fields: ${JSON.stringify(timestampsToSort, null, 2)}`
+    );
+  }
 
-        // filter out the search results that match with the values found in the list.
-        // the resulting set are signals to be indexed, given they are not duplicates
-        // of signals already present in the signals index.
-        const filteredEvents = await filterEventsAgainstList({
-          listClient,
-          exceptionsList,
-          logger,
-          eventSearchResult: searchResult,
-          buildRuleMessage,
+  for (const timestamp of timestamps) {
+    const totalToFromTuples = getSignalTimeTuples({
+      logger,
+      ruleParamsFrom: ruleParams.from,
+      ruleParamsTo: ruleParams.to,
+      ruleParamsMaxSignals: ruleParams.maxSignals,
+      gap,
+      previousStartedAt,
+      interval,
+      buildRuleMessage,
+    });
+    logger.debug(buildRuleMessage(`totalToFromTuples: ${totalToFromTuples.length}`));
+
+    while (totalToFromTuples.length > 0) {
+      const tuple = totalToFromTuples.pop();
+      if (tuple == null || tuple.to == null || tuple.from == null) {
+        logger.error(buildRuleMessage(`[-] malformed date tuple`));
+        return createSearchAfterReturnType({
+          success: false,
+          errors: ['malformed date tuple'],
         });
+      }
+      signalsCreatedCount = 0;
```

Comment:
Note: we could end up with 2*max signals per time window now, since we query the default and override timestamps separately.